src/web/charconfig_controller.cpp

```diff
@@ -0,0 +1,162 @@
+#include "http.hpp"
+#include "sqllock.hpp"
+#include "utils.hpp"
+#include "web.hpp"
+
+HANDLER_FUNC(charconfig_save) {
+	if (!isAuthorized(req, false)) {
+		res.status = 400;
+		res.set_content("Error", "text/plain");
+		return;
+	}
+	
+	auto account_id = std::stoi(req.get_file_value("AID").content);
+	auto char_id = std::stoi(req.get_file_value("GID").content);
+	auto world_name_str = req.get_file_value("WorldName").content;
+	auto world_name = world_name_str.c_str();
+	std::string data;
+
+	if (req.has_file("data")) {
+		data = req.get_file_value("data").content;
+		addToJsonObject(data, "\"Type\": 1");
+	} else {
+		data = "{\"Type\": 1}";
+	}
+	SQLLock sl(WEB_SQL_LOCK);
+	sl.lock();
+	auto handle = sl.getHandle();
+	SqlStmt * stmt = SqlStmt_Malloc(handle);
+	if (SQL_SUCCESS != SqlStmt_Prepare(stmt,
+			"SELECT `account_id` FROM `%s` WHERE (`account_id` = ? AND `char_id` = ? AND `world_name` = ?) LIMIT 1",
+			char_configs_table)
+		|| SQL_SUCCESS != SqlStmt_BindParam(stmt, 0, SQLDT_INT, &account_id, sizeof(account_id))
+		|| SQL_SUCCESS != SqlStmt_BindParam(stmt, 1, SQLDT_INT, &char_id, sizeof(char_id))
+		|| SQL_SUCCESS != SqlStmt_BindParam(stmt, 2, SQLDT_STRING, (void *)world_name, strlen(world_name))
+		|| SQL_SUCCESS != SqlStmt_Execute(stmt)
+	) {
+		SqlStmt_ShowDebug(stmt);
+		SqlStmt_Free(stmt);
+		sl.unlock();
+		res.status = 400;
+		res.set_content("Error", "text/plain");
+		return;
+	}
+
+	if (SqlStmt_NumRows(stmt) <= 0) {
+		if (SQL_SUCCESS != SqlStmt_Prepare(stmt,
+				"INSERT INTO `%s` (`account_id`, `char_id`, `world_name`, `data`) VALUES (?, ?, ?, ?)",
+				char_configs_table)
+			|| SQL_SUCCESS != SqlStmt_BindParam(stmt, 0, SQLDT_INT, &account_id, sizeof(account_id))
+			|| SQL_SUCCESS != SqlStmt_BindParam(stmt, 1, SQLDT_INT, &char_id, sizeof(char_id))
+			|| SQL_SUCCESS != SqlStmt_BindParam(stmt, 2, SQLDT_STRING, (void *)world_name, strlen(world_name))
+			|| SQL_SUCCESS != SqlStmt_BindParam(stmt, 3, SQLDT_STRING, (void *)data.c_str(), strlen(data.c_str()))
+			|| SQL_SUCCESS != SqlStmt_Execute(stmt)
+		) {
+			SqlStmt_ShowDebug(stmt);
+			SqlStmt_Free(stmt);
+			sl.unlock();
+			res.status = 400;
+			res.set_content("Error", "text/plain");
+			return;
+		}
+	} else {
+		if (SQL_SUCCESS != SqlStmt_Prepare(stmt,
+				"UPDATE `%s` SET `data` = ? WHERE (`account_id` = ? AND `char_id` = ? AND `world_name` = ?)",
+				char_configs_table)
+			|| SQL_SUCCESS != SqlStmt_BindParam(stmt, 0, SQLDT_STRING, (void *)data.c_str(), strlen(data.c_str()))
+			|| SQL_SUCCESS != SqlStmt_BindParam(stmt, 1, SQLDT_INT, &account_id, sizeof(account_id))
+			|| SQL_SUCCESS != SqlStmt_BindParam(stmt, 2, SQLDT_INT, &char_id, sizeof(char_id))
+			|| SQL_SUCCESS != SqlStmt_BindParam(stmt, 3, SQLDT_STRING, (void *)world_name, strlen(world_name))
+			|| SQL_SUCCESS != SqlStmt_Execute(stmt)
+		) {
+			SqlStmt_ShowDebug(stmt);
+			SqlStmt_Free(stmt);
+			sl.unlock();
+			res.status = 400;
+			res.set_content("Error", "text/plain");
+			return;
+		}
+	}
```

Comment:
Just a stupid thought: REPLACE INTO?

Reply:
I think `INSERT INTO ... ON DUPLICATE KEY UPDATE ...` is the preferred method now, I'll use that.

Reply:
Actually, at some point, the client only sends part of the config, if some parts haven't changed.
Look at pandas: https://github.com/PandasWS/Pandas/blob/d7c5cca3ca7eabd6add1cf082e1bf98f3f724788/src/web/charconfig_controller.cpp#L214-L222

We can add support for this later, but for now I'd like to keep a select & insert/update.